File: frontend/lib/models/auction.dart
import 'package:intl/intl.dart';
import 'package:lance_certo/models/auction_status.dart';
import 'package:lance_certo/models/product.dart';
import 'package:lance_certo/models/user.dart';

class Auction {
  Auction({
    required this.startDateAndTime,
    required this.endDateAndTime,
    required this.initialPrice,
    required this.minimunBidIncrement,
    this.product,
    this.seller,
    this.status,
    this.id,
    this.currentBid,
    this.currentBidder,
    this.winner,
    this.createdAt,
    this.updatedAt,
  });

  final int? id;
  final Product? product;
  final User? seller;
  final DateTime startDateAndTime;
  final DateTime endDateAndTime;
  final double initialPrice;
  final double minimunBidIncrement;
  double? currentBid;
  User? currentBidder;
  AuctionStatus? status;
  User? winner;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  factory Auction.fromJson(Map<String, dynamic> json) {
    return Auction(
      id: json['id'] as int?,
      product: Product.fromJson(json['product'] as Map<String, dynamic>),
      seller: User.fromJson(json['product']['seller'] as Map<String, dynamic>),
      startDateAndTime: DateTime.parse(json['startDateAndTime'] as String),
      endDateAndTime: DateTime.parse(json['endDateAndTime'] as String),
      initialPrice: (json['initialPrice'] as num).toDouble(),
      minimunBidIncrement: (json['minimunBidIncrement'] as num).toDouble(),
      currentBid: (json['currentBid'] as num?)?.toDouble(),
      currentBidder: json['currentBidder'] != null
          ? User.fromJson(json['currentBidder'] as Map<String, dynamic>)
          : null,
      status: AuctionStatus.fromString(json['status'] as String),
      winner: json['winner'] != null
          ? User.fromJson(json['winner'] as Map<String, dynamic>)
          : null,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'] as String)
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
    'productId': product!.id,
    'startDateAndTime': startDateAndTime.toIso8601String(),
    'endDateAndTime': endDateAndTime.toIso8601String(),
    'initialPrice': initialPrice,
    'minimunBidIncrement': minimunBidIncrement,
  };

  @override
  String toString() {
    return 'Auction(\n'
        '  id: $id,\n'
        '  product: ${product?.toString() ?? 'N/A'},\n'
        '  seller: ${seller?.username ?? 'N/A'},\n'
        '  startDateAndTime: ${DateFormat('dd/MM/yyyy HH:mm').format(startDateAndTime)},\n'
        '  endDateAndTime: ${DateFormat('dd/MM/yyyy HH:mm').format(endDateAndTime)},\n'
        '  initialPrice: $initialPrice,\n'
        '  minimunBidIncrement: $minimunBidIncrement,\n'
        '  currentBid: ${currentBid ?? 'N/A'},\n'
        '  currentBidder: ${currentBidder?.username ?? 'N/A'},\n'
        '  status: ${status?.name ?? 'N/A'},\n'
        '  winner: ${winner?.username ?? 'N/A'},\n'
        '  createdAt: ${createdAt != null ? DateFormat('dd/MM/yyyy HH:mm').format(createdAt!) : 'N/A'},\n'
        '  updatedAt: ${updatedAt != null ? DateFormat('dd/MM/yyyy HH:mm').format(updatedAt!) : 'N/A'},\n'
        ')';
  }
}

File: backend/src/main/java/br/com/hematsu/lance_certo/model/Auction.java
package br.com.hematsu.lance_certo.model;

import java.io.Serializable;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;

import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OrderBy;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@NoArgsConstructor
@Entity(name = "tb_auctions")
public class Auction implements Serializable {
  private static final long serialVersionUID = 2405172041950251807L;

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  private Long id;

  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "product_id", nullable = false)
  private Product product;

  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "seller_id", nullable = false)
  private User seller;

  @Column(nullable = false)
  private LocalDateTime startDateAndTime;

  @Column(nullable = false)
  private LocalDateTime endDateAndTime;

  @Column(nullable = false, precision = 19, scale = 2)
  private BigDecimal initialPrice;

  @Column(nullable = false, precision = 19, scale = 2)
  private BigDecimal minimunBidIncrement;

  @Column(nullable = false, precision = 19, scale = 2)
  private BigDecimal currentBid;

  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "current_bidder_id")
  private User currentBidder;

  @Enumerated(EnumType.STRING)
  @Column(nullable = false)
  private AuctionStatus status;

  @ManyToOne(fetch = FetchType.LAZY)
  @JoinColumn(name = "winner_id")
  private User winner;

  @CreationTimestamp
  @Column(nullable = false)
  private LocalDateTime createdAt;

  @UpdateTimestamp
  @Column(nullable = false)
  private LocalDateTime updatedAt;

  @OneToMany(mappedBy = "auction", cascade = CascadeType.ALL, orphanRemoval = true)
  @OrderBy("amount Desc, createdAt ASC")
  private List<Bid> bids;
}
